solidity/contracts/lib/zeto_base.sol

```diff
@@ -62,14 +81,39 @@ abstract contract ZetoBase is IZetoBase, ZetoCommon {
             if (i > 0 && sortedInputs[i] == sortedInputs[i - 1]) {
                 revert UTXODuplicate(sortedInputs[i]);
             }
-            if (_utxos[sortedInputs[i]] == UTXOStatus.UNKNOWN) {
+            if (
+                _lockedUtxos[sortedInputs[i]] == UTXOStatus.UNKNOWN &&
+                _utxos[sortedInputs[i]] == UTXOStatus.UNKNOWN
+            ) {
                 revert UTXONotMinted(sortedInputs[i]);
-            } else if (_utxos[sortedInputs[i]] == UTXOStatus.SPENT) {
+            }
+            if (
+                (inputsLocked &&
+                    _lockedUtxos[sortedInputs[i]] == UTXOStatus.SPENT) ||
+                (!inputsLocked && _utxos[sortedInputs[i]] == UTXOStatus.SPENT)
+            ) {
                 revert UTXOAlreadySpent(sortedInputs[i]);
             }
+            if (
+                !inputsLocked &&
+                _lockedUtxos[sortedInputs[i]] == UTXOStatus.UNSPENT
+            ) {
+                revert UTXOAlreadyLocked(sortedInputs[i]);
+            }
+            if (
```

Comment:
Feels like all these "if" statements could be nested into a big `if (inputsLocked) {} else {}` block. It's almost entirely different checks for locked inputs vs unlocked ones.